packages/mobile/src/identity/commentEncryption.ts

```diff
@@ -0,0 +1,309 @@
+  // they don't benefit from comment encryption
+  const fromKey: Buffer | null = yield call(getCommentKey, fromAddress)
+  if (!fromKey) {
+    Logger.debug(TAG + 'encryptComment', 'No sender key found, skipping encryption')
+    return comment
+  }
+
+  const toKey: Buffer | null = yield call(getCommentKey, toAddress)
+  if (!toKey) {
+    Logger.debug(TAG + 'encryptComment', 'No recipient key found, skipping encryption')
+    return comment
+  }
+
+  let commentToEncrypt = comment
+  if (features.PHONE_NUM_METADATA_IN_TRANSFERS && includePhoneNumMetadata) {
+    Logger.debug(TAG + 'encryptComment', 'Including phone number metadata in comment')
+    const selfPhoneDetails: PhoneNumberHashDetails | undefined = yield call(
+      getUserSelfPhoneHashDetails
+    )
+    commentToEncrypt = embedPhoneNumberMetadata(comment, selfPhoneDetails)
+  }
+
+  const { comment: encryptedComment, success } = encryptCommentRaw(commentToEncrypt, toKey, fromKey)
+
+  if (success) {
+    Logger.debug(TAG + 'encryptComment', 'Encryption succeeded')
+    return encryptedComment
+  } else {
+    Logger.error(TAG + 'encryptComment', 'Encrytion failed, returning raw comment')
+    return comment
+  }
+}
+
+interface DecryptedComment {
+  comment: string | null
+  e164Number?: string
+  salt?: string
+}
+
+// Memoize to avoid computing decryptions more than once per comment
+// TODO investigate whether its worth it to save this in persisted state, maybe Apollo cache?
+export const decryptComment = memoize(_decryptComment, (...args) => values(args).join('_'))
+
+function _decryptComment(
+  comment: string | null,
+  commentKeyPrivate: string | null,
+  isSender: boolean
+): DecryptedComment {
+  Logger.debug(TAG + 'decryptComment', 'Decrypting comment')
+
+  if (!features.USE_COMMENT_ENCRYPTION || !comment || !commentKeyPrivate) {
+    Logger.debug(TAG + 'decryptComment', 'Invalid params, skipping decryption')
+    return { comment }
+  }
+
+  const { comment: decryptedComment, success } = decryptCommentRaw(
+    comment,
+    hexToBuffer(commentKeyPrivate),
+    isSender
+  )
+
+  if (success) {
+    Logger.debug(TAG + 'decryptComment', 'Comment decryption succeeded')
+    return extractPhoneNumberMetadata(decryptedComment)
+  } else {
+    Logger.error(TAG + 'decryptComment', 'Decrypting comment failed, returning raw comment')
+    return { comment }
+  }
+}
+
+export function embedPhoneNumberMetadata(
+  comment: string,
+  phoneNumberDetails?: PhoneNumberHashDetails
+) {
+  return phoneNumberDetails
+    ? comment + METADATA_CONTENT_SEPARATOR + phoneNumberDetails.e164Number + phoneNumberDetails.salt
+    : comment
+}
+
+export function extractPhoneNumberMetadata(commentData: string) {
+  const phoneNumMetadata = commentData.match(PHONE_METADATA_REGEX)
+  if (!phoneNumMetadata || phoneNumMetadata.length < 4) {
+    return { comment: commentData }
+  }
+
+  return {
+    comment: phoneNumMetadata[1],
```

Comment:
I'm not super knowledgable about regex, but would the comment not be index 0 here?

Reply:
also wondering how this solution would scale if we add more metadata and perhaps some of it is missing.  feels almost like we should send json or some type:value in the metadata for more foolproof parsing

Reply:
No, index 0 would be the full string. 
This scheme would not work well for generalized metadata but we can cross that bridge when we come to it :)